File: app/normalize.py
"""Turn the portal's loose data into clean, typed records."""
from collections import Counter, defaultdict
from datetime import datetime

LEVELS = ["zone", "circle", "division", "subdivision", "substation", "feeder", "dt"]


def _num(v) -> float | None:
    if v in ("", None):
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


def readings(raw: list[dict]) -> list[dict]:
    """Portal rows -> ISO timestamps, numeric fields, no duplicate timestamps.

    The portal occasionally emits a second row for the same timestamp with blank
    kwh/voltR. We keep the first, non-blank row per timestamp.
    """
    by_ts: dict[datetime, dict] = {}
    for r in raw:
        try:
            ts = datetime.strptime(r["timestamp"], "%d/%m/%Y %H:%M")
        except (KeyError, ValueError):
            continue
        row = {"timestamp": ts.isoformat(), "kwh": _num(r.get("kwh")),
               "kvah": _num(r.get("kvah")), "voltR": _num(r.get("voltR"))}
        if ts not in by_ts or (by_ts[ts]["kwh"] is None and row["kwh"] is not None):
            by_ts[ts] = row
    out = [by_ts[k] for k in sorted(by_ts)]
    if len(out) >= 2:
        step = (datetime.fromisoformat(out[1]["timestamp"]) - datetime.fromisoformat(out[0]["timestamp"])).total_seconds()
        gran = "30min" if step == 1800 else "daily" if step == 86400 else f"{int(step)}s"
    else:
        gran = "unknown"
    return out, gran  # type: ignore[return-value]


def daily_consumption(rows: list[dict]) -> list[dict]:
    """kWh consumed per calendar day from a cumulative register."""
    by_day: dict[str, list[dict]] = defaultdict(list)
    for r in rows:
        if r["kwh"] is not None:
            by_day[r["timestamp"][:10]].append(r)
    days = sorted(by_day)
    out = []
    prev_last = None
    for d in days:
        first, last = by_day[d][0]["kwh"], by_day[d][-1]["kwh"]
        base = prev_last if prev_last is not None else first
        out.append({"date": d, "kwh": round(last - base, 3), "readings": len(by_day[d])})
        prev_last = last
    return out


def clean_meters(export: list[dict], transformers: list[dict]) -> tuple[list[dict], list[dict]]:
    """Repair hierarchy gaps and return (meters, issues).

    Observed problems in the export: a node with a name but blank code (or the
    reverse), and a blank feeder even though the DT list knows its feeder.
    We fill gaps from other meters that share the same name/code, then from
    the transformer list, and record every repair as an issue.
    """
    name_to_code: dict[str, dict[str, str]] = {l: {} for l in LEVELS}
    code_to_name: dict[str, dict[str, str]] = {l: {} for l in LEVELS}
    name_votes: dict[str, dict[str, Counter]] = {l: defaultdict(Counter) for l in LEVELS}
    for m in export:
        for l in LEVELS:
            n = m["hierarchy"].get(l) or {}
            if n.get("code") and n.get("name"):
                name_to_code[l].setdefault(n["name"], n["code"])
                name_votes[l][n["code"]][n["name"]] += 1
    for l in LEVELS:
        code_to_name[l] = {c: v.most_common(1)[0][0] for c, v in name_votes[l].items()}
    dt_feeder = {t["code"]: t["feederCode"] for t in transformers}

    meters, issues = [], []
    for m in export:
        h = {}
        for l in LEVELS:
            n = dict(m["hierarchy"].get(l) or {})
            code, name = n.get("code") or "", n.get("name") or ""
            if not code and name and name in name_to_code[l]:
                code = name_to_code[l][name]
                issues.append({"meterId": m["meterId"], "level": l, "fix": "code_from_name", "value": code})
            if code and not name and code in code_to_name[l]:
                name = code_to_name[l][code]
                issues.append({"meterId": m["meterId"], "level": l, "fix": "name_from_code", "value": name})
            if l == "feeder" and not code and m.get("dtCode") in dt_feeder:
                code = dt_feeder[m["dtCode"]]
                name = code_to_name[l].get(code, "")
                issues.append({"meterId": m["meterId"], "level": l, "fix": "feeder_from_dt", "value": code})
            if l == "dt" and code and name != code_to_name[l].get(code, name):
                issues.append({"meterId": m["meterId"], "level": l, "fix": "dt_name_alias",
                               "value": f"{name} -> {code_to_name[l][code]}"})
                name = code_to_name[l][code]
            h[l] = {"code": code or None, "name": name or None}
        geo = m.get("geo") or {}
        meters.append({
            "meterId": m["meterId"], "serialNo": m["serialNo"], "make": m["make"],
            "phaseType": m["phaseType"], "installStatus": m["installStatus"],
            "installType": m["installType"], "build": m["build"], "dtCode": m["dtCode"],
            "hierarchy": h,
            "location": {"lat": _num(geo.get("lat")), "lng": _num(geo.get("lng"))},
        })
    return meters, issues


def build_tree(meters: list[dict]) -> dict:
    """Nested zone > ... > dt tree with meter counts.

    Codes are reused under different parents (e.g. D-01 sits under C-01, C-03
    and C-05), so a node's identity is its full path, not its code.
    """
    root: dict = {"children": {}, "meterCount": 0}
    for m in meters:
        node = root
        node["meterCount"] += 1
        for l in LEVELS:
            n = m["hierarchy"][l]
            key = n["code"] or f"?{n['name'] or 'unknown'}"
            child = node["children"].setdefault(key, {"level": l, "code": n["code"], "name": n["name"],
                                                       "children": {}, "meterCount": 0})
            child["meterCount"] += 1
            node = child

    def to_list(n: dict) -> dict:
        kids = [to_list(c) for c in n["children"].values()]
        out = {k: v for k, v in n.items() if k != "children"}
        out["children"] = sorted(kids, key=lambda c: c["code"] or "~")
        return out

    return to_list(root)

File: src/normalize.ts
// Pure functions that turn the portal's loose data into clean records.
import type { ExportMeter, RawReading, Transformer } from "./portal.ts";

export const LEVELS = ["zone", "circle", "division", "subdivision", "substation", "feeder", "dt"] as const;
export type Level = (typeof LEVELS)[number];

export type HierarchyNode = { code: string | null; name: string | null };
export type Meter = Omit<ExportMeter, "hierarchy" | "geo"> & {
  hierarchy: Record<Level, HierarchyNode>;
  location: { lat: number; lng: number } | null;
};
export type Issue = { meterId: string; level: Level; fix: "code_from_name" | "name_from_code" | "feeder_from_dt" | "dt_name_alias"; value: string };
export type Reading = { timestamp: string; kwh: number | null; kvah: number | null; voltR: number | null };
export type TreeNode = { level: Level | "root"; code: string | null; name: string | null; meterCount: number; children: TreeNode[] };

const toNumber = (v: string): number | null => (v === "" || Number.isNaN(Number(v)) ? null : Number(v));

// "30/06/2026 00:30" -> "2026-06-30T00:30:00"
function toIso(ts: string): string | null {
  const m = /^(\d{2})\/(\d{2})\/(\d{4}) (\d{2}):(\d{2})$/.exec(ts);
  return m ? `${m[3]}-${m[2]}-${m[1]}T${m[4]}:${m[5]}:00` : null;
}

// Typed readings, sorted, one row per timestamp. The portal sometimes emits a
// second row for the same timestamp with blank kwh; the row with a value wins.
export function cleanReadings(raw: RawReading[]): { granularity: string; data: Reading[] } {
  const byTime = new Map<string, Reading>();
  for (const r of raw) {
    const timestamp = toIso(r.timestamp);
    if (!timestamp) continue;
    const row = { timestamp, kwh: toNumber(r.kwh), kvah: toNumber(r.kvah), voltR: toNumber(r.voltR) };
    const existing = byTime.get(timestamp);
    if (!existing || (existing.kwh === null && row.kwh !== null)) byTime.set(timestamp, row);
  }
  const data = [...byTime.values()].sort((a, b) => a.timestamp.localeCompare(b.timestamp));
  const [first, second] = data;
  const stepMin = first && second ? (Date.parse(second.timestamp) - Date.parse(first.timestamp)) / 60_000 : 0;
  const granularity = stepMin === 30 ? "30min" : stepMin === 1440 ? "daily" : "unknown";
  return { granularity, data };
}

// kWh used per calendar day from a cumulative register: the day's last reading
// minus the previous day's last reading (or the day's first, for day one).
export function dailyConsumption(readings: Reading[]): { date: string; kwh: number; readings: number }[] {
  const days = new Map<string, number[]>();
  for (const r of readings) {
    if (r.kwh === null) continue;
    const date = r.timestamp.slice(0, 10);
    days.set(date, [...(days.get(date) ?? []), r.kwh]);
  }
  const out: { date: string; kwh: number; readings: number }[] = [];
  let previousClose: number | null = null;
  for (const [date, values] of [...days].sort(([a], [b]) => a.localeCompare(b))) {
    const open: number = previousClose ?? values[0] ?? 0;
    const close: number = values[values.length - 1] ?? open;
    out.push({ date, kwh: Math.round((close - open) * 1000) / 1000, readings: values.length });
    previousClose = close;
  }
  return out;
}

// Repairs hierarchy gaps seen in the export: a node with a name but no code
// (or the reverse), a blank feeder the DT list can fill, and a DT with two
// names. Every repair is recorded so consumers can audit it.
export function cleanMeters(exported: ExportMeter[], transformers: Transformer[]): { meters: Meter[]; issues: Issue[] } {
  const codeByName = new Map<string, string>(); // key: "level|name"
  const nameVotes = new Map<string, Map<string, number>>(); // key: "level|code"
  for (const m of exported) {
    for (const level of LEVELS) {
      const { code, name } = m.hierarchy[level];
      if (!code || !name) continue;
      if (!codeByName.has(`${level}|${name}`)) codeByName.set(`${level}|${name}`, code);
      const votes = nameVotes.get(`${level}|${code}`) ?? new Map<string, number>();
      votes.set(name, (votes.get(name) ?? 0) + 1);
      nameVotes.set(`${level}|${code}`, votes);
    }
  }
  const majorityName = (level: Level, code: string): string | undefined =>
    [...(nameVotes.get(`${level}|${code}`) ?? [])].sort((a, b) => b[1] - a[1])[0]?.[0];
  const feederByDt = new Map(transformers.map((t) => [t.code, t.feederCode]));

  const issues: Issue[] = [];
  const meters = exported.map((m): Meter => {
    const fixNode = (level: Level): HierarchyNode => {
      let { code, name } = m.hierarchy[level];
      const note = (fix: Issue["fix"], value: string) => issues.push({ meterId: m.meterId, level, fix, value });
      if (!code && name) {
        code = codeByName.get(`${level}|${name}`) ?? "";
        if (code) note("code_from_name", code);
      }
      if (level === "feeder" && !code) {
        code = feederByDt.get(m.dtCode) ?? "";
        if (code) note("feeder_from_dt", code);
      }
      const canonical = code ? majorityName(level, code) : undefined;
      if (code && !name && canonical) {
        name = canonical;
        note("name_from_code", name);
      }
      if (level === "dt" && canonical && name !== canonical) {
        note("dt_name_alias", `${name} -> ${canonical}`);
        name = canonical;
      }
      return { code: code || null, name: name || null };
    };
    const { hierarchy: _h, geo, ...nameplate } = m;
    return {
      ...nameplate,
      hierarchy: {
        zone: fixNode("zone"), circle: fixNode("circle"), division: fixNode("division"),
        subdivision: fixNode("subdivision"), substation: fixNode("substation"),
        feeder: fixNode("feeder"), dt: fixNode("dt"),
      },
      location: geo,
    };
  });
  return { meters, issues };
}

// Nested zone > ... > DT tree with meter counts. Codes are reused under
// different parents (D-01 sits under C-01, C-03 and C-05), so a node is
// identified by its path, not its code.
export function buildTree(meters: Meter[]): TreeNode {
  const root: TreeNode = { level: "root", code: null, name: null, meterCount: 0, children: [] };
  for (const m of meters) {
    let node = root;
    node.meterCount++;
    for (const level of LEVELS) {
      const { code, name } = m.hierarchy[level];
      let child = node.children.find((c) => c.code === code && (code !== null || c.name === name));
      if (!child) {
        child = { level, code, name, meterCount: 0, children: [] };
        node.children.push(child);
        node.children.sort((a, b) => (a.code ?? "~").localeCompare(b.code ?? "~"));
      }
      child.meterCount++;
      node = child;
    }
  }
  return root;
}

export function distanceKm(a: { lat: number; lng: number }, b: { lat: number; lng: number }): number {
  const rad = Math.PI / 180;
  const h =
    0.5 - Math.cos((b.lat - a.lat) * rad) / 2 +
    (Math.cos(a.lat * rad) * Math.cos(b.lat * rad) * (1 - Math.cos((b.lng - a.lng) * rad))) / 2;
  return 12742 * Math.asin(Math.sqrt(h));
}
